simulate/simulate.cc

```diff
@@ -0,0 +1,1885 @@
+        if (selbody>=0) {
+          mju_copy3(simulate->cam.lookat, selpnt);
+        }
+
+        // switch to tracking camera if dynamic body clicked
+        if (selmode==3 && selbody>0) {
+          // mujoco camera
+          simulate->cam.type = mjCAMERA_TRACKING;
+          simulate->cam.trackbodyid = selbody;
+          simulate->cam.fixedcamid = -1;
+
+          // UI camera
+          simulate->camera = 1;
+          mjui_update(SECT_RENDERING, -1, &simulate->ui0, &simulate->uistate, &simulate->con);
+        }
+      }
+
+      // set body selection
+      else {
+        if (selbody>=0) {
+          // record selection
+          simulate->pert.select = selbody;
+          simulate->pert.skinselect = selskin;
+
+          // compute localpos
+          mjtNum tmp[3];
+          mju_sub3(tmp, selpnt, simulate->d->xpos+3*simulate->pert.select);
+          mju_mulMatTVec(simulate->pert.localpos, simulate->d->xmat+9*simulate->pert.select, tmp, 3, 3);
+        } else {
+          simulate->pert.select = 0;
+          simulate->pert.skinselect = -1;
+        }
+      }
+
+      // stop perturbation on select
+      simulate->pert.active = 0;
+    }
+
+    return;
+  }
+
+  // 3D release
+  if (state->type==mjEVENT_RELEASE && state->dragrect==3 && simulate->m) {
+    // stop perturbation
+    simulate->pert.active = 0;
+
+    return;
+  }
+
+  // 3D move
+  if (state->type==mjEVENT_MOVE && state->dragrect==3 && simulate->m) {
+    // determine action based on mouse button
+    mjtMouse action;
+    if (state->right) {
+      action = state->shift ? mjMOUSE_MOVE_H : mjMOUSE_MOVE_V;
+    } else if (state->left) {
+      action = state->shift ? mjMOUSE_ROTATE_H : mjMOUSE_ROTATE_V;
+    } else {
+      action = mjMOUSE_ZOOM;
+    }
+
+    // move perturb or camera
+    mjrRect r = state->rect[3];
+    if (simulate->pert.active)
+      mjv_movePerturb(simulate->m, simulate->d, action, state->dx/r.height, -state->dy/r.height,
+                      &simulate->scn, &simulate->pert);
+    else
+      mjv_moveCamera(simulate->m, action, state->dx/r.height, -state->dy/r.height,
+                     &simulate->scn, &simulate->cam);
+
+    return;
+  }
+}
+
+void uiRender(mjuiState* state) {
+  mj::Simulate* simulate = (mj::Simulate*)(state->userdata);
+  simulate->render();
+}
+
+// drop file callback
+void drop(mj::Simulate* simulate, int count, const char** paths) {
+  // make sure list is non-empty
+  if (count>0) {
+    mju::strcpy_arr(simulate->dropfilename, paths[0]);
+    simulate->droploadrequest = 1;
+  }
+}
+
+void uiDrop(mjuiState* state, int count, const char** paths) {
+  mj::Simulate* simulate = (mj::Simulate*)(state->userdata);
+  drop(simulate, count, paths);
+}
+
+} // end unnamed namespace
+
+namespace mujoco {
+namespace mju = ::mujoco::sample_util;
+
+//---------------------------------- init -------------------------------------------------
+
+// create object and initialize the simulate ui
+Simulate::Simulate(void) {
```

Comment:
I think that you need to zero init `ui0` and `ui1` here.

Reply:
Are you thinking about writing swift bindings for this shared library? Python bindings were straightforward, see [here](https://github.com/aftersomemath/mujoco/blob/simulate-python4/python/mujoco/simulate.cc) and [here](https://github.com/aftersomemath/mujoco/blob/simulate-python4/python/mujoco/simulate.py).

I see that you have recently transcribed `simulate` [to swift](https://github.com/liuliu/swift-mujoco/blob/main/Examples/simulate/main.swift) so maybe that is your preferred approach?

Reply:
Transcribe `simulate` to Swift is a way for me to validate my binding generation, it has been successful at shaking off quite a few bugs :)

Swift binding prefers to work with C interface. It is a bit different than Python in that regard. Since I am now done with the translation, I will just follow this PR in spirit with a pure `final class Simulate` in Swift then. It should be pretty straightforward now.